[Parser][SR-711] Couple "expected declaration" Diagnose With A Note

dduan opened the pull request:
As reported in [SR-711](https://bugs.swift.org/browse/SR-711), when an (unexpected) statement appears in a type declaration, we note the beginning of the declaration in addiction to the existing diagnostic.

dduan commented:
@jrose-apple I updated the diagnostic message to match the bug report.

slavapestov commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,28 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
+          diagnose(ty->getLoc(), diag::note_in_decl_extension, false,
+                   ty->getNameStr());
+        } else if (auto ed = dyn_cast<ExtensionDecl>(CurDeclContext)) {
+          if (auto extension = ed->getExtendedType()) {
```
This whole branch is dead because getAsNominalTypeOrNominalTypeExtensionContext() returns the extended type if we have an extension -- you should just move the ```isa<ExtensionDecl>()``` inside the ```if (ty) {```.

dduan replied:
I had the suspension. Thanks for confirming it :)

slavapestov commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,28 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
         }
       }
+      if (CurDeclContext) {
+        auto ty =
```
This returns a decl, not a type. ```auto *decl``` would be better.

slavapestov commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,28 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
+            diagnose(ed->getLoc(), diag::note_in_decl_extension, true,
+                     StringRef(extension->getString()));
+          } else {
+            diagnose(ed->getLoc(), diag::note_in_extension);
```
This shouldn't happen at all

dduan replied:
`getAsNominalTypeOrNominalTypeExtensionContext()` would not return a valid type for extensions (`base=<null>` from dumping the context). My guess is this happens during the first pass?

slavapestov commented on test/Parse/recovery.swift:
```diff
@@ -461,7 +461,7 @@ func use_BracesInsideNominalDecl1() {
 //===--- Recovery for wrong decl introducer keyword.
 
-class WrongDeclIntroducerKeyword1 {
+class WrongDeclIntroducerKeyword1 { // expected-note{{in declaration of 'WrongDeclIntroducerKeyword1':}}
```
Add tests for extensions here

dduan commented:
@slavapestov updated.

dduan commented:
@slavapestov I added a separate test for this change: https://github.com/dduan/swift/blob/SR771_stmt_in_type_body_2/test/Parse/recovery.swift#L462

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,23 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
+          CurDeclContext->getAsNominalTypeOrNominalTypeExtensionContext();
+        if (nominal) {
+          diagnose(nominal->getLoc(), diag::note_in_decl_extension, false,
+                   nominal->getNameStr());
```
Better to use `getName()` here.

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,23 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
       }
+      if (CurDeclContext) {
+        auto nominal =
+          CurDeclContext->getAsNominalTypeOrNominalTypeExtensionContext();
```
This won't work for extensions. We're still in parsing; we haven't resolved any names yet. If you want to do names for extensions, you should try to get common patterns out of `getExtendedTypeLoc`; otherwise, you can just dyn_cast to NominalTypeDecl.

jrose-apple commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2005,13 +2005,23 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
+        auto nominal =
+          CurDeclContext->getAsNominalTypeOrNominalTypeExtensionContext();
+        if (nominal) {
+          diagnose(nominal->getLoc(), diag::note_in_decl_extension, false,
```
We can't put notes before the primary warning/error; they'll be associated with the previous diagnostic instead of the next one. Just put it afterwards, it's fine.

dduan committed:
[Parser] Couple "expected declaration" Diagnose With A Note

As reported in SR-711, when an (unexpected) statement appears in a type
declaration, we note the beginning of the declaration in addiction to the
existing diagnostic.

dduan committed:
[Parser] updated tests for "expected declaration" companion note

dduan commented:
@jrose-apple Updated:

* `getName()` replaces `getNameStr()`
* extension name is now part of the note, as a consequence, number of `NOTE` added to *.def* is reduced to 1
* note now gets issued after the error.

jrose-apple commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -143,6 +143,9 @@ WARNING(lex_editor_placeholder_in_playground,none,
 //------------------------------------------------------------------------------
 
+NOTE(note_in_decl_extension,none,
+     "in %select{declaration|extension}0 of %1:", (bool, Identifier))
```
Nitpick: please drop the colon, since it's not coming before the warning anymore.

jrose-apple commented on test/Parse/recovery.swift:
```diff
@@ -459,9 +459,18 @@ func use_BracesInsideNominalDecl1() {
+}
+
+extension SR771 { // expected-note {{in extension of 'SR771':}}
+    print("The room where it happened, the room where it happened") // expected-error {{expected declaration}}
```
Bonus points awarded for your choice of cultural reference. :-)

dduan replied:
I'm still trying to sneak this one in:
![372](https://cloud.githubusercontent.com/assets/75067/13186026/ec0731d8-d6f8-11e5-9d19-93865613692a.jpg)

😀

dduan commented:
@jrose-apple ok the ':' is no more.

dduan commented:
@jrose-apple BTW, do you think it'd be beneficial to add a `getExtendedTypeName()` method to `ExtensionDecl`, using the logic in this PR?

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Eh, until there's more use of it I think `getExtendedTypeName` isn't too interesting—once we get through early stages of type-checking, the extended type should be filled in. Thanks, Daniel!

dduan commented:
Test failure seems unrelated?

dduan commented:
Looking through recent CI build failures for other PRs, lots of them fail for the same reason. I think it's safe to assume it's okay 😁

jrose-apple commented:
The failure is an existing issue in the package manager. I think that's good enough.

dduan commented:
Thank you!

- Daniel Duan




On Mon, Feb 22, 2016 at 1:40 PM -0800, "Jordan Rose" <notifications@github.com> wrote:












Merged #1282.



—
Reply to this email directly or view it on GitHub.

jrose-apple commented:
Feel free to mark the JIRA issue as resolved.
